tests/unittests/test_gcp_authorization.py

```diff
@@ -0,0 +1,125 @@
+from typing import Optional
+
+import pytest
+from unittest.mock import Mock, patch
+from google.auth.credentials import Credentials
+from google.auth.exceptions import RefreshError
+
+from ai21.clients.vertex.gcp_authorization import GCPAuthorization
+from google.auth.transport.requests import Request
+
+_TEST_PROJECT_ID = "test-project"
+
+
+@pytest.mark.parametrize(
+    ids=[
+        "when_project_id_not_found__should_raise_error",
+        "when_project_id_found__should_set_auth",
+    ],
+    argvalues=[(None,), (_TEST_PROJECT_ID,)],
+    argnames=["project_id"],
+)
+@patch("google.auth.default")
+@patch.object(GCPAuthorization, GCPAuthorization._get_gcp_request.__name__)
+def test__gcp_authorization_set_auth(
+    mock_gcp_request: Mock,
+    mock_default: Mock,
+    mock_gcp_credentials: Mock,
+    project_id: Optional[str],
+):
+    mock_default.return_value = (mock_gcp_credentials, project_id)
+    mock_gcp_request.return_value = Mock(spec=Request)
+
+    auth = GCPAuthorization(project_id=project_id)
+
+    if project_id is None:
+        with pytest.raises(ValueError):
+            auth._set_auth()
+    else:
+        auth._set_auth()
+        assert auth._credentials == mock_gcp_credentials
+        assert auth.project_id == _TEST_PROJECT_ID
+
+
+@patch("google.auth.default")
+@patch.object(GCPAuthorization, GCPAuthorization._get_gcp_request.__name__)
+def test__gcp_authorization_set_auth__project_id_mismatch__should_raise_error(
+    mock_gcp_request: Mock,
+    mock_default: Mock,
+    mock_gcp_credentials: Mock,
+):
+    mock_default.return_value = (mock_gcp_credentials, "another-project")
+    mock_gcp_request.return_value = Mock(spec=Request)
+
+    auth = GCPAuthorization(project_id=_TEST_PROJECT_ID)
+
+    with pytest.raises(ValueError):
```

Comment:
Since this class raises a lot of different errors, I think it'd be good practice to assert the exception message